benchbuild/utils/revision_ranges.py

```diff
@@ -0,0 +1,276 @@
+
+    Args:
+        bad_commits: A list of `bad` commits.
+        good_commits: A list of `good` commits.
+        comment: See :func:`AbstractRevisionRange.comment()`.
+    """
+    def __init__(self,
+                 bad_commits: tp.List[str],
+                 good_commits: tp.List[str],
+                 comment: tp.Optional[str] = None):
+        super().__init__(comment)
+        self.__bad_commit_ids = bad_commits
+        self.__good_commit_ids = good_commits
+        # cache for commit hashes
+        self.__revision_list: tp.Optional[tp.List[str]] = None
+
+    def init_cache(self, repo_path: Path) -> None:
+        self.__revision_list = []
+        repo = pygit2.Repository(str(repo_path))
+        git = _get_git_for_path(repo_path)
+
+        bad_commits = [repo.get(bug_id) for bug_id in self.__bad_commit_ids]
+        good_commits = [repo.get(fix_id) for fix_id in self.__good_commit_ids]
+
+        # start search from all branch heads
+        heads = git("show-ref", "--heads", "-s").strip().split("\n")
+        for head in heads:
+            self.__revision_list.extend([
+                str(commit.id) for commit in _find_blocked_commits(
+                    repo.get(head), good_commits, bad_commits)
+            ])
+
+    def __iter__(self) -> tp.Iterator[str]:
+        if self.__revision_list is None:
+            raise AssertionError
+        return self.__revision_list.__iter__()
+
+
+def block_revisions(blocks: tp.List[AbstractRevisionRange]) -> tp.Any:
+    """
+    Decorator for project classes for blacklisting/blocking revisions.
+
+    ATTENTION: This decorator depends on things introduced by the
+    @with_git decorator and therefore must be used above that decorator.
+
+    This adds a new static method ``is_blocked_revision`` that checks
+    whether a given revision id is marked as blocked.
+
+    Args:
+        blocks: A list of :class:`AbstractRevisionRange` s.
+    """
+    def revision_blocker_decorator(cls: tp.Any) -> tp.Any:
+        def is_blocked_revision_impl(
+                rev_id: str) -> tp.Tuple[bool, tp.Optional[str]]:
+            """
+            Checks whether a revision is blocked or not. Also returns the
+            reason for the block if available.
+            """
+            # trigger caching for BlockedRevisionRanges
+            if not cls.blocked_revisions_initialized:
+                cls.blocked_revisions_initialized = True
+                for block in blocks:
+                    # TODO: how to access local repository?
+                    block.init_cache(cls.repo_path())
+
+            for b_entry in blocks:
+                for b_item in b_entry:
+                    if b_item.startswith(rev_id):
+                        return True, b_entry.comment
+            return False, None
+
+        cls.blocked_revisions_initialized = False
+        cls.is_blocked_revision = is_blocked_revision_impl
+        return cls
+
+    return revision_blocker_decorator
```

Comment:
Do we want this on BB level? Not sure if blocking revisions is useful for BB.
@simbuerg

Reply:
That is totally under the control of a project. We can always rename/generalize it later.